[BUGFIX] fix jsonapi error handling when not using jquery

aclarembeau opened the pull request:
Fixes #6936

snewcomer commented on packages/-ember-data/tests/integration/adapter/handle-response-test.js:
```diff
@@ -175,4 +175,35 @@ module('integration/adapter/handle-response', function(hooks) {
+    class TestAdapter extends JSONAPIAdapter {
+      handleResponse(status, headers, payload, requestData) {
+        handleResponseCalled++;
+        assert.deepEqual(payload, errorObject);
```
Could add a failure message - 

`assert.deepEqual(payload, errorObject, 'payload from handleResponse matches expected error');`

aclarembeau replied:
Done :)

snewcomer commented on packages/-ember-data/tests/integration/adapter/handle-response-test.js:
```diff
@@ -175,4 +175,35 @@ module('integration/adapter/handle-response', function(hooks) {
     assert.equal(handleResponseCalled, 1, 'handle response is called');
   });
+
+  test('handleResponse is called with correct parameters on string repsonse with 422 status', async function(assert) {
```
`assert.expect(3);`

snewcomer replied:
r/repsonse/response

aclarembeau replied:
I don't get the assert.expect point, could you tell me more?

snewcomer replied:
Yeah!  So because you have assertions inside of stubbed methods that don't have to be called for the test to pass, by adding an `assert.expect`, we ensure that is is called.

snewcomer commented on packages/adapter/addon/rest.js:
```diff
@@ -1301,9 +1301,11 @@ function fetchSuccessHandler(adapter, payload, response, requestData) {
   if (responseData.status === 200 && payload instanceof Error) {
     responseData.errorThrown = payload;
-    payload = responseData.errorThrown.payload;
+    payload = responseData.errorThrown;
```
Do you think we need this reassignment to `payload`?

aclarembeau replied:
Indeed, that was an error on my side. 
I shouldn't even change anything in that branch. 
My fix was to handle the case where payload is a javascript string representing a JSONAPI error object. In that branch, payload is an Error object, so, I shouldn't change anything here.

snewcomer commented on packages/adapter/addon/rest.js:
```diff
@@ -1301,9 +1301,11 @@ function fetchSuccessHandler(adapter, payload, response, requestData) {
   let responseData = fetchResponseData(response);
+  payload = adapter.parseErrorResponse(payload);
+
   if (responseData.status === 200 && payload instanceof Error) {
```
It is interesting the error you throw in your test is 422, but the check here is 200.  I thought there was a reason for this but I can't remember why.  Do you happen to know?

aclarembeau replied:
After re-reding the code, it seems that there could be some case where the HTTP status is 200, but an error is thrown (for example due to a runtime error). In that case, we fall into the status=200 branch, so we don't need to parse payload there.

aclarembeau commented:
@snewcomer , I've updated my code with: 
- better respect of linting rules 
- updated logic (not parsing JSON payload if payload is not JSON, but an error object)

efx commented on packages/adapter/addon/rest.js:
```diff
@@ -1301,11 +1301,13 @@ function fetchSuccessHandler(adapter, payload, response, requestData) {
 
 function fetchErrorHandler(adapter, payload, response, errorThrown, requestData) {
   let responseData = fetchResponseData(response);
+
```
@aclarembeau total nitpick, but can you remove this whitespace?

aclarembeau replied:
Sure, it's done :)

efx commented:
The asset size checks are failing by 28 bytes; not sure if that amount should hold up this PR.
@runspired thoughts?

runspired commented:
Asset size change is totally fine. I agree there is a bug and that the solution is to JSON.parse something.

What I am not convinced of is that this is the right place in the process to do this. To be convinced I need to spend a little time (or Igor needs to spend a little time) walking through the fetch flow and confirming where the point of divergence in behavior occurs so that we are confident we're fixing it at the source and not much later in the flow.

runspired commented on packages/adapter/addon/rest.js:
```diff
@@ -1301,11 +1301,13 @@ function fetchSuccessHandler(adapter, payload, response, requestData) {
     payload = responseData.errorThrown.payload;
   } else {
     responseData.errorThrown = errorThrown;
+    payload = adapter.parseErrorResponse(payload);
```
A few comments after reading through the existing implementation:

I believe the necessary change is either this OR

```
  if (!response.ok) {
    return payload;
  }
```

^ this line in determineBodyPromise should possibly attempt JSON.parse. **I lean heavily towards the former (current PR) solution**

The primary difference is in whether we give the adapter the chance to do the parsing. Note the Ajax codepath calls this for all Ajax errors (I'm not sure if a 404 triggers an error though, as odd as that sounds at first glance, it doesn't trigger one in fetch for instance).

But if it does...

```
let payload = adapter.parseErrorResponse(jqXHR.responseText);
```

^ this is what would be doing the similar parsing for the Ajax branch.

cc @igorT
